builders/reservationBuilder.go

```diff
@@ -0,0 +1,205 @@
+)
+
+// ReservationBuilder is a struct that can build reservations
+type ReservationBuilder struct {
+	Reservation workloads.Reservation
+	Bcdb        *client.Client
+	Mainui      *tfexplorer.UserIdentity
+	assets      []string
+	seedPath    string
+	dryRun      bool
+}
+
+// NewReservationBuilder creates a new ReservationBuilder
+func NewReservationBuilder(bcdb *client.Client, mainui *tfexplorer.UserIdentity) *ReservationBuilder {
+	reservation := workloads.Reservation{}
+	reservation.Epoch = schema.Date{Time: time.Now()}
+	return &ReservationBuilder{
+		Reservation: reservation,
+		Bcdb:        bcdb,
+		Mainui:      mainui,
+	}
+}
+
+// LoadReservationBuilder loads a reservation builder based on a file path
+func LoadReservationBuilder(reader io.Reader) (*ReservationBuilder, error) {
+	reservation := workloads.Reservation{}
+	err := json.NewDecoder(reader).Decode(&reservation)
+	if err != nil {
+		return &ReservationBuilder{}, err
+	}
+
+	return &ReservationBuilder{Reservation: reservation}, nil
+}
+
+// Save saves the reservation builder to an IO.Writer
+func (r *ReservationBuilder) Save(writer io.Writer) error {
+	err := json.NewEncoder(writer).Encode(r.Reservation)
+	if err != nil {
+		return err
+	}
+	return err
+}
+
+// Build returns the reservation
+func (r *ReservationBuilder) Build() workloads.Reservation {
+	return r.Reservation
+}
+
+// Deploy deploys the reservation
+func (r *ReservationBuilder) Deploy() (wrklds.ReservationCreateResponse, error) {
+	userID := int64(r.Mainui.ThreebotID)
+	signer, err := client.NewSigner(r.Mainui.Key().PrivateKey.Seed())
+	if err != nil {
+		return wrklds.ReservationCreateResponse{}, errors.Wrapf(err, "could not find seed file at %s", r.seedPath)
+	}
+
+	r.Reservation.CustomerTid = userID
+	// we always allow user to delete his own reservations
+	r.Reservation.DataReservation.SigningRequestDelete.QuorumMin = 1
+	r.Reservation.DataReservation.SigningRequestDelete.Signers = []int64{userID}
+
+	// set allowed the currencies as provided by the user
+	r.Reservation.DataReservation.Currencies = r.assets
+
+	bytes, err := json.Marshal(r.Reservation.DataReservation)
+	if err != nil {
+		return wrklds.ReservationCreateResponse{}, err
+	}
+
+	r.Reservation.Json = string(bytes)
+	_, signature, err := signer.SignHex(r.Reservation.Json)
+	if err != nil {
+		return wrklds.ReservationCreateResponse{}, errors.Wrap(err, "failed to sign the reservation")
+	}
+
+	r.Reservation.CustomerSignature = signature
+
+	if r.dryRun {
+		enc := json.NewEncoder(os.Stdout)
+		enc.SetIndent("", "  ")
+		return wrklds.ReservationCreateResponse{}, enc.Encode(r.Reservation)
+	}
+
+	response, err := r.Bcdb.Workloads.Create(r.Reservation)
+	if err != nil {
+		return wrklds.ReservationCreateResponse{}, errors.Wrap(err, "failed to send reservation")
+	}
+
+	return response, nil
+}
+
+// DeleteReservation deletes a reservation by id
+func (r *ReservationBuilder) DeleteReservation(resID int64) error {
```

Comment:
I'm not sure this belongs to the builder itself. This is more about what to do with the reservation then how to build it.

Reply:
Where do you suggest we move this logic? Back to the tfuser provision command?

Reply:
Well I think we need another part of the library that deal with this.
I think the library needs to have 2 packages. On for the builders and one for all the logic regarding registering reservations, cancelling etc...

Reply:
Yes, i agree with Chris, the builder is concerned with building a valid reservation objects, validate and save/load them. And the we have another high level client (in the same module but not necessary the same package) that can deploy/delete and such.